kernel/src/log_segment.rs

```diff
@@ -112,13 +269,137 @@ impl LogSegment {
     }
 }
 
+/// Returns a fallible iterator of [`ParsedLogPath`] that are between the provided `start_version` (inclusive)
+/// and `end_version` (inclusive). [`ParsedLogPath`] may be a commit or a checkpoint.  If `start_version` is
+/// not specified, the files will begin from version number 0. If `end_version` is not specified, files up to
+/// the most recent version will be included.
+///
+/// Note: this calls [`FileSystemClient::list_from`] to get the list of log files.
+fn parsed_log_files_iter(
+    fs_client: &dyn FileSystemClient,
+    log_root: &Url,
+    start_version: impl Into<Option<Version>>,
+    end_version: impl Into<Option<Version>>,
+) -> DeltaResult<impl Iterator<Item = DeltaResult<ParsedLogPath>>> {
+    let start_version = start_version.into().unwrap_or(0);
+    let end_version = end_version.into();
+
+    let version_prefix = format!("{:020}", start_version);
+    let start_from = log_root.join(&version_prefix)?;
+
+    // Check that the provided version is less than or equal to the end version if it exists
+    let lte_end_version = move |version: u64| {
+        end_version.is_none() || end_version.is_some_and(|end_version| version <= end_version)
+    };
+
+    Ok(fs_client
+        .list_from(&start_from)?
+        .map(|meta| ParsedLogPath::try_from(meta?))
+        // TODO this filters out .crc files etc which start with "." - how do we want to use these kind of files?
+        .filter_map_ok(identity)
+        .take_while(move |x| match x {
+            Ok(x) => lte_end_version(x.version),
+            Err(_) => true,
+        }))
```

Comment:
maybe something like this is simpler? also maybe flag that Err and None are filtered out
```suggestion
    let lte_end_version = move |version: u64| {
        end_version.map_or(true, |end_version| version <= end_version)
    };

    Ok(fs_client.list_from(&start_from)?
        // TODO this filters out .crc files etc which start with "." - how do we want to use these kind of files?
        .filter_map(|meta| ParsedLogPath::try_from(meta.ok()?).ok())
        .take_while(move |log_path| lte_end_version(log_path.version)))
```

Reply:
Hmm is it a good idea to be filtering out errors though?

Reply:
new discussion below (https://github.com/delta-incubator/delta-kernel-rs/pull/495#discussion_r1843771435) so im resolving this